### PR TITLE
docs(todos): create tutorial

### PR DESCRIPTION
**Pull request by felangel**
## Status

**READY**

## Description

- docs(todos): create tutorial

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [X] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: docs/docs/tutorials/todos.md [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+:::info
+**Difficulty**: 🟠 Intermediate<br/>
+**Length**: 30 minutes
+
+Before getting started, [read the prerequisites](/docs/overview#prerequisites) to make sure your development environment is ready.
+:::
+
+## Overview
+
+In this tutorial, we're going to build an app that exposes a two endpoints which allow us to perform C.R.U.D operations on a list of todos.
```
Comment (erickzanardo) [inline docs/docs/tutorials/todos.md:17]:
```suggestion
In this tutorial, we're going to build an app that exposes two endpoints which allow us to perform C.R.U.D operations on a list of todos.
```

File: docs/docs/tutorials/todos.md [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+## Todos Data Source
+
+### Creating `package:todos_data_source`
+
+Now that we have a running application, we need to define an interface for a todos data source which will be responsible for exposing APIs to perform C.R.U.D operations on a list of todos.
+
+Since the todos data source is not tightly coupled to our Dart Frog application, we can create it as a package.
+
+:::tip
+Decomposing a project into one or more packages is a form of modularization which can help with maintainability and reusability.
```
Comment (erickzanardo) [inline docs/docs/tutorials/todos.md:119]: 💙

File: docs/docs/tutorials/todos.md [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+You should see detailed request logs in the console due to the `requestLogger` middleware that look similar to:
+
+```bash
+2022-08-09T17:43:35.816387  0:00:00.016484 GET     [200] /todos
+2022-08-09T17:44:05.561021  0:00:00.022465 POST    [201] /todos
+```
+
+:::
+
+🎉 Congrats, you've created an `todos` application using Dart Frog. View the [full source code](https://github.com/VeryGoodOpenSource/dart_frog/tree/main/examples/todos).
```
Comment (erickzanardo) [inline docs/docs/tutorials/todos.md:665]:
```suggestion
🎉 Congrats, you've created a `todos` application using Dart Frog. View the [full source code](https://github.com/VeryGoodOpenSource/dart_frog/tree/main/examples/todos).
```

**Review by erickzanardo — APPROVED** [review]

File: docs/docs/tutorials/todos.md [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+  --url http://localhost:8080
+```
+
+If everything succeeded, you should see `Welcome to Dart Frog!`.
+
+## Todos Data Source
+
+### Creating `package:todos_data_source`
+
+Now that we have a running application, we need to define an interface for a todos data source which will be responsible for exposing APIs to perform C.R.U.D operations on a list of todos.
```
Comment (scarletteliza) [inline docs/docs/tutorials/todos.md:114]:
```suggestion
Now that we have a running application, we need to define an interface for a todos data source which will be responsible for exposing APIs to perform C.R.U.D operations on a list of todos.
```
is interface the best word here?

File: docs/docs/tutorials/todos.md (hunk 2 above)
Comment (scarletteliza) [reply docs/docs/tutorials/todos.md:119]: should we link to anything here @felangel ??

File: docs/docs/tutorials/todos.md [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+  /// Converts this [Todo] into a `Map<String, dynamic>`.
+  Map<String, dynamic> toJson() => _$TodoToJson(this);
+
+  @override
+  List<Object?> get props => [id, title, description, isCompleted];
+}
+```
+
+:::info
+The `Todo` class uses [package:json_serializable](https://pub.dev/packages/json_serializable) to handle generating the code to (de)serialize to/from JSON.
```
Comment (scarletteliza) [inline docs/docs/tutorials/todos.md:280]:
```suggestion
The `Todo` class uses [package:json_serializable](https://pub.dev/packages/json_serializable) to handle generating the code to (de)serialize to and from JSON.
```

File: docs/docs/tutorials/todos.md [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+The `Todo` class extends uses [package:equatable](https://pub.dev/packages/equatable) to override `==` and `hashCode` so that we can compare `Todo` instances by value.
+:::
+
+Next, we need to use [package:build_runner](https://pub.dev/packages/build_runner) to generate the relevant code for `json_serializable`:
+
+```bash
+dart run build_runner build --delete-conflicting-outputs
+```
+
+We should see that the `todos.g.dart` file was generated and our code should not have any errors/warnings at this point.
```
Comment (scarletteliza) [inline docs/docs/tutorials/todos.md:291]:
```suggestion
We should see that the `todos.g.dart` file was generated and our code should not have any errors or warnings at this point.
```
stop it with the slashes :P

File: docs/docs/tutorials/todos.md [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+  Future<Todo> update(String id, Todo todo) async {
+    return _cache.update(id, (value) => todo);
+  }
+
+  @override
+  Future<void> delete(String id) async => _cache.remove(id);
+}
+```
+
+That's it! We're done with the data sources for our Dart Frog application and we're ready to start working on the Dart Frog app itself!
```
Comment (scarletteliza) [inline docs/docs/tutorials/todos.md:430]:
```suggestion
That's it! We're done making the data sources for our Dart Frog application and we're ready to start working on the Dart Frog app itself!
```
I feel like we needed a verb here so I added "making" but feel free to switch out with another

**Review by scarletteliza — APPROVED** [review]

File: docs/docs/tutorials/todos.md (hunk 4 above)
Comment (felangel) [reply docs/docs/tutorials/todos.md:114]: Yeah I think so

File: docs/docs/tutorials/todos.md (hunk 2 above)
Comment (felangel) [reply docs/docs/tutorials/todos.md:119]: I'll try to link to an existing blog post tomorrow 👍

File: docs/docs/tutorials/todos.md [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+:::info
+**Difficulty**: 🟠 Intermediate<br/>
+**Length**: 30 minutes
+
+Before getting started, [read the prerequisites](/docs/overview#prerequisites) to make sure your development environment is ready.
+:::
+
+## Overview
+
+In this tutorial, we're going to build an app that exposes two endpoints which allow us to perform C.R.U.D operations on a list of todos.
```
Comment (wolfenrain) [inline docs/docs/tutorials/todos.md:17]: Never seen CRUD being referenced like that, with the dots.

I always use CRUD, as defined by [wiki](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete)

File: docs/docs/tutorials/todos.md [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+
+## Creating a new app
+
+To create a new Dart Frog app, open your terminal, `cd` into the directory where you'd like to create the app, and run the following command:
+
+```bash
+dart_frog create todos
+```
+
+You should see an output similar to:
```
Comment (wolfenrain) [inline docs/docs/tutorials/todos.md:64]: 👍

File: docs/docs/tutorials/todos.md (hunk 4 above)
Comment (wolfenrain) [reply docs/docs/tutorials/todos.md:114]: Shouldn't abstraction be a better fit? We are defining an abstraction first and then the implementation of that abstraction.

**Review by wolfenrain — APPROVED** [review]

File: docs/docs/tutorials/todos.md (hunk 4 above)
Comment (felangel) [reply docs/docs/tutorials/todos.md:114]: Good call, I like abstraction better as well 👍

File: docs/docs/tutorials/todos.md (hunk 8 above)
Comment (felangel) [reply docs/docs/tutorials/todos.md:17]: Good point, will replace

**Pull request merged**